pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,491 @@
+	for _, ipPool := range ipPools {
+		nodeSel, _ := metav1.LabelSelectorAsSelector(&ipPool.Spec.NodeSelector)
+		nodeList, _ := c.nodeLister.List(nodeSel)
+		for _, node := range nodeList {
+			nodeName := node.Name
+			if !nodeSet.Has(nodeName) {
+				nodeSet.Insert(node.Name)
+				nodes = append(nodes, node)
+			}
+		}
+	}
+	return
+}
+
+func (c *Cluster) syncClusterMembers() (clusterNodes []string) {
+	nodes, err := c.filterNodes()
+	if err != nil {
+		klog.ErrorS(err, "Listing Nodes failed")
+		return
+	}
+	klog.V(3).InfoS("Sync cluster members", "node num", len(nodes))
+
+	for _, node := range nodes {
+		nodeAddr, err := k8s.GetNodeAddr(node)
+		if err != nil {
+			klog.ErrorS(err, "Obtain local IP address from K8s node failed")
+			continue
+		}
+		member := fmt.Sprintf("%s:%d", nodeAddr, c.bindPort)
+		clusterNodes = append(clusterNodes, member)
+	}
+	return
+}
+
+func (c *Cluster) shouldJoinCluster(nodeName string) bool {
+	node, err := c.nodeLister.Get(nodeName)
+	if err != nil {
+		klog.ErrorS(err, "Get node failed", "nodeName", nodeName)
+		return false
+	}
+	return c.matchNodeSelectors(node)
+}
+
+func (c *Cluster) matchNodeSelectors(node *corev1.Node) bool {
+	ipPools, err := c.ipPoolLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "matchNodeSelectors failed")
+		return false
+	}
+	for _, ipPool := range ipPools {
+		nodeSelector, _ := metav1.LabelSelectorAsSelector(&ipPool.Spec.NodeSelector)
+		if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *Cluster) memberNum() int {
+	c.memberRWLock.RLock()
+	defer c.memberRWLock.RUnlock()
+	return c.existingMembers.Len()
+}
+
+func (c *Cluster) addExistingMembers(members ...string) {
+	c.memberRWLock.Lock()
+	defer c.memberRWLock.Unlock()
+	c.existingMembers.Insert(members...)
+	klog.V(2).InfoS("Add existingMember", "memberNum", c.existingMembers.Len())
+}
+
+func (c *Cluster) updateExistingMember(node *corev1.Node, add bool) {
+	c.memberRWLock.Lock()
+	defer c.memberRWLock.Unlock()
+	nodeAddr, err := k8s.GetNodeAddr(node)
+	if err != nil {
+		klog.ErrorS(err, "Failed to obtain local IP address")
+		return
+	}
+	member := fmt.Sprintf("%s:%d", nodeAddr, c.bindPort)
+	oldNum := c.existingMembers.Len()
+	if add {
+		c.existingMembers.Insert(member)
+		newNum := c.existingMembers.Len()
+		klog.V(2).Infof("Add new member (%s) in local existing memberlist, num %d->%d", member, oldNum, newNum)
+	} else {
+		if !c.existingMembers.Has(member) {
+			return
+		}
+		c.existingMembers.Delete(member)
+		newNum := c.existingMembers.Len()
+		klog.V(2).Infof("Delete member (%s) in local existing memberlist, num %d->%d", member, oldNum, newNum)
+	}
+	if c.nodeName == node.Name {
+		c.updateLocalNodeStatus()
+	}
+}
+
+func (c *Cluster) joinOrLeaveCluster(join bool) {
+	if join {
+		c.memberRWLock.RLock()
+		defer c.memberRWLock.RUnlock()
+		clusterNodes := c.existingMembers.List()
+		n, err := c.mList.Join(clusterNodes)
```

Comment:
question: When a remote Node leaves the cluster (because of no pool selects it), will this Node remove it from the members it will try to connect? or once a node joined, it will be in the cluster unless recreated?

Reply:
> question: When a remote Node leaves the cluster (because of no pool selects it), will this Node remove it from the members it will try to connect? or once a node joined, it will be in the cluster unless recreated?

"will this Node remove it from the members" means local node remove remote node from cluster?

The cluster node can only leave itself or failed, local node can not remove remote node from cluster.